runtime/src/main/java/org/corfudb/runtime/collections/PersistedStreamingMap.java

```diff
@@ -0,0 +1,320 @@
+                                 @NonNull Options options,
+                                 @NonNull ISerializer serializer,
+                                 @NonNull CorfuRuntime corfuRuntime) {
+        try {
+            RocksDB.destroyDB(dataPath.toFile().getAbsolutePath(), options);
+            this.rocksDb = RocksDB.open(options, dataPath.toFile().getAbsolutePath());
+        } catch (RocksDBException e) {
+            throw new UnrecoverableCorfuError(e);
+        }
+        this.serializer = serializer;
+        this.corfuRuntime = corfuRuntime;
+        openedResources.add(this);
+    }
+
+    /**
+     * A Java compatible {@link RocksIterator} implementation
+     */
+    public class RocksDbIterator implements Iterator<Entry<K, V>> {
+        private RocksIterator iterator;
+        private Entry<K, V> current;
+        private Entry<K, V> next;
+
+        RocksDbIterator(RocksIterator iterator) {
+            this.iterator = iterator;
+        }
+
+        /**
+         * Ensure that this iterator is operating under the correct assumptions.
+         */
+        private void checkInvariants() {
+            // RocksDB does not support multi-threaded access.
+            // If the iterator was created by some thread, it also has to be consumed by it.
+            if (!iterator.isOwningHandle()) {
+                throw new IllegalStateException("Detected multi-threaded access to this iterator.");
+            }
+
+            try {
+                iterator.status();
+            } catch (RocksDBException e) {
+                throw new UnrecoverableCorfuError(
+                        "There was an error reading the persisted map.", e);
+            }
+        }
+
+        /**
+         * {@inheritDoc}
+         */
+        @Override
+        public boolean hasNext() {
+            // If we have the next element pipelined, go ahead and return true.
+            if (next != null) {
+                return true;
+            }
+
+            // If the iterator is valid, this means that the next entry exists.
+            checkInvariants();
+            if (iterator.isValid()) {
+                // Go ahead and cache that entry.
+                next = new AbstractMap.SimpleEntry(
+                        serializer.deserialize(Unpooled.wrappedBuffer(iterator.key()), corfuRuntime),
+                        serializer.deserialize(Unpooled.wrappedBuffer(iterator.value()), corfuRuntime));
+                // Advance the underlying iterator.
+                iterator.next();
+            } else {
+                // If there is no more elements to consume, we should release the resources.
+                iterator.close();
+            }
+
+            return next != null;
+        }
+
+        /**
+         * {@inheritDoc}
+         */
+        @Override
+        public Entry<K, V> next() {
+            if (hasNext()) {
+                current = next;
+                next = null;
```

Comment:
I don't get the motivation for the `next` variable?
I see that we can get rid of `next` variable completely